DIPs/DIPxxxx_attributes.md

```diff
@@ -0,0 +1,185 @@
+
+### Links
+
+[Forum discussion](https://forum.dlang.org/thread/wnddmlmfinqqfccdlhqc@forum.dlang.org)
+
+## Terminology
+
+Groups of attributes that are mutually exclusive (such as `@safe`, `@system`, `@trusted`) and presence/absence attributes
+(e.g. `pure`, `nothrow`) and their (currently) non-existant logical negations are called _attribute groups_.
+
+Attributes from `core.attribute` are called core attributes.
+
+## Rationale
+
+Many users feel that the default attributes have the wrong defaults and given that attributes are not invertable,  putting 
+```
+pure: nothrow: @nogc: @safe:
+```
+at the top of a file means that one can never "undo" those attributes.
+This DIP proposes a solution to easily change the default values of the attributes while also allowing symbol specific 
+overrides of individual attribute groups.
+
+## Description
+
+Move all (DMD) compiler recognized attributes into `core.attribute`, making them symbols in their own right, 
+grouping by attribute groups into `enum`s, each with
+* a value `inferred`. The compiler shall determine the value of the attribute. It is illegal to have a function declaration `inferred`.
+* the attribute(s)
+* (if applicable) the attributes' logical negation.
+
+It is suggested that 
+* `inferred` have the value 0, and the compiler shall be required to replace enum core attributes
+that have the value `inferred` to the acual value as determined by the compiler, i.e. no symbols in object files or in reflection shall
+have a core attribute with the value `inferred`. 
+* the current default have the value 1.
+* the negation of the default (`@safe` for `@safe`/`@system`/`@trusted`) the value 2.
+* any other values (e.g. `@trusted`) continue from the value 3.
+
+A module declaration may be tagged with zero or more attribute groups, to apply to all symbols (bar templates which remain inferred with explicit tagging) declared within the module acting as the default.
+If any attribute groups are absent, then the value for that attribute group default to the corresponding value in `core.attribute.defaultAttributeSet`, which will have the values of the current defauls, but may be versioned in druntime as the end user wishes, or with command line switches (e.g. `-safe` or if Type_Info / Module Info generation is added as an attribute `-betterC`).
+
+As all the attributes are now symbols we can group the in an `AliasSeq` like fashion to apply them en masse, as is done in LDC for [`@fastmath`](https://github.com/ldc-developers/druntime/blob/ldc/src/ldc/attributes.d#L58).
+
+It is illegal to explicitly provide more than one attribute from any given attribute group as they are mutually exclusive. 
+Attributes applied explicity to any symbol override the module default attribute set.
+
+### Attributes & attribute like compiler behaviour encompassed in this DIP
+
+Encompassed:
+
+* pure
+* @nothrow
+* @nogc
+* @safe/@system/@trusted
+
+Optionally encompassed:
+
+* final
```

Comment:
> I don't think return should be defaultable. I think the main issue with a scope default for parameters is that there's no way to enable it currently at module scope, so maybe it's outside the scope (heh) of this DIP.

From what I gather from DIP1000, the compiler is supposed to infer `scope`/`return` for parameters, so I'm not sure how great the benefits would be. Having said that it would be nice to be able to determine what the compiler should be inferring.

> Some people will probably ask for immutable to be defaultable (again not sure how practical this is). 

That is whole different can of worms and likely to cause much more and is not in the scope of this DIP.

>  Is there a reason why keyword attributes can't work?

Well that would require a fundamental semantic change (tuples must be symbols or values, i.e what does `mySet[2]` evaluate to? ), whereas converting them to enums fits with the current semantics and solves the same problem, albeit at the cost of deprecatable breaking changes.